p2p: Prevent integer overflow in LocalServiceInfo::nScore

codeabysss opened the pull request:
The overflow for signed arithmetic yields undefined behavior.
This changes prevents undefined behavior in local address scoring by saturating `nScore` updates at `INT_MAX` in both `SeenLocal()` and `AddLocal()` update paths.

Fixes: #24049.

Crypt-iQ commented on src/net.h:
```diff
@@ -102,6 +98,9 @@ static constexpr bool DEFAULT_V2_TRANSPORT{true};
 typedef int64_t NodeId;
 
+/** Get the score of a local address. */
+int GetnScore(const CService& addr);
```
I want to hear what others think, but I'm not a fan of changing the function signature for it to only be used in a test. I think the test and this change can go, and you can provide a small diff (that doesn't get merged) to apply on top of this PR so reviewers can see that the issue is fixed.

sedited replied:
I think making an existing function public for testing some behaviour is fine, but agree that providing a small diff to test the fix would be ok too. This is a small patch, so maybe dropping is easier and less code to review. @Crypt-iQ can you also take another look at its current state?

Crypt-iQ replied:
Ok I will take another look

l0rinc replied:
Alternatively we could add a test-local lambda like:
```C++
BOOST_AUTO_TEST_CASE(LocalAddress_nScore_Overflow)
{
    g_reachable_nets.Add(NET_IPV4);
    CService addr{UtilBuildAddress(0x002, 0x001, 0x001, 0x001), 1000}; // 2.1.1.1:1000
    auto check_score{[&](int expected) {
        const auto local_addresses{m_node.connman->getNetLocalAddresses()};
        const auto it{local_addresses.find(addr)};
        BOOST_REQUIRE(it != local_addresses.end());
        BOOST_CHECK_EQUAL(it->second.nScore, expected);
    }};

    const int initial_score = 1000;
    BOOST_REQUIRE(AddLocal(addr, initial_score));
    BOOST_REQUIRE(IsLocal(addr));
    check_score(initial_score);

...
```

Crypt-iQ replied:
I'm ~0 on a test being added if we're doing a `SaturatedAdd` in `SeenLocal` & `AddLocal`, then we're just testing `SaturatedAdd`?. Using `m_node.connman->getNetLocalAddresses` does have the benefit of not needing to change the function signature though.

l0rinc replied:
> Using m_node.connman->getNetLocalAddresses does have the benefit of not needing to change the function signature though.

Yes, that was the reason for suggesting it.

> then we're just testing SaturatedAdd

We're testing the current behavior black-box style to document it and avoid regressing later.

Crypt-iQ replied:
> We're testing the current behavior black-box style to document it and avoid regressing later.

Sure, I just think it's a bit overkill for what could be a ~3-line change. Hopefully somebody would see saturation and not remove it, but maybe that's too much to ask and not robust. I don't really mind at this point (hence the ~0), the issue has been open for over 3 years with ~4-5 PRs trying to fix what should have been an extremely simple fix.

Crypt-iQ reviewed (commented):
A couple comments:

It seems like some of the wording in the OP was copied from the other PR #33072 and I think it would be better if you put things in your own words (my personal preference). Discussion in the other PR also mentioned that this has little to do with memory efficiency since it's just 4 bytes at the end of the day. I think the commit message could also be cleaned up a bit with a short title and then a descriptive paragraph underneath.

Crypt-iQ commented on src/test/net_tests.cpp:
```diff
@@ -800,6 +800,36 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+    BOOST_REQUIRE(IsLocal(addr));
+    BOOST_CHECK_EQUAL(GetnScore(addr), initial_score);
+
+    // SeenLocal increments the score
```
Can move the above "SeenLocal increments the nScore is below max" comment to here

codeabysss replied:
this is done

Crypt-iQ commented on src/test/net_tests.cpp:
```diff
@@ -800,6 +800,36 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+    BOOST_CHECK(SeenLocal(addr));
+    BOOST_CHECK_EQUAL(GetnScore(addr), initial_score + 1);
+
+    // SeenLocal saturates at max
```
maybe "Set nScore to max"?

codeabysss replied:
this is done

Crypt-iQ commented on src/test/net_tests.cpp:
```diff
@@ -800,6 +800,36 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+    BOOST_REQUIRE(AddLocal(addr, std::numeric_limits<int>::max()));
+    BOOST_CHECK_EQUAL(GetnScore(addr), std::numeric_limits<int>::max());
+
+    // a couple increments should saturate
```
maybe "Check that nScore is saturated"?

Crypt-iQ replied:
Also, could remove the for loop and do the increment once instead.

codeabysss replied:
this is done

Crypt-iQ reviewed (commented):
Left some nits about the test comments. I think because the functional tests use the local interface and there is a `IsRoutable` check (which itself checks that the address isn't local) before calling `SeenLocal`, this would be hard or impossible to check in a functional test. So I think changing the function signature is fine.

hebasto commented:
Please rebase once more to resolve the CI issue.

Crypt-iQ commented on src/test/net_tests.cpp:
```diff
@@ -38,6 +38,15 @@ using namespace std::literals;
 using util::ToString;
 
+namespace {
+int GetLocalScore(const CService& addr)
```
This does work, but I think changing the function signature like you did originally was fine. If there's any change to `GetnScore` then this function would have to be updated to reflect that, which could be forgotten.

luke-jr replied:
I agree, no reason to just duplicate the code

codeabysss replied:
reverted back thanks

Crypt-iQ commented:
ACK 33103d5c4fe2c0df85d1d8acf81d21846df6ee54

Crypt-iQ commented:
I think the `sock_tests` are flaking and unrelated.

sedited reviewed (approved):
ACK 33103d5c4fe2c0df85d1d8acf81d21846df6ee54

I'm not sure how useful the test is here as per https://github.com/bitcoin/bitcoin/pull/34028#discussion_r3109321797 . I think it would be better to drop it.

Crypt-iQ commented:
> I'm not sure how useful the test is here as per https://github.com/bitcoin/bitcoin/pull/34028#discussion_r3109321797 . I think it would be better to drop it.

+1. It was good to verify the fix works when running with ubsan, but I think it can be dropped and the function signature reverted back to normal.

l0rinc commented on src/net.cpp:
```diff
@@ -325,7 +325,9 @@ bool SeenLocal(const CService& addr)
     const auto it = mapLocalHost.find(addr);
     if (it == mapLocalHost.end()) return false;
-    ++it->second.nScore;
+    if (it->second.nScore < std::numeric_limits<int>::max()) {
```
We already have https://github.com/bitcoin/bitcoin/blob/8b49e2dd4eed93d08a3d9681d444aaf441ab0037/src/util/overflow.h#L45-L60 for this exact usecase, let's use it here

l0rinc commented on src/test/net_tests.cpp:
```diff
@@ -802,6 +802,33 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+    CService addr{UtilBuildAddress(0x002, 0x001, 0x001, 0x001), 1000}; // 2.1.1.1:1000
+
+    const int initial_score = 1000;
+    BOOST_REQUIRE(AddLocal(addr, initial_score));
```
We should probably cover the `nScore` update in https://github.com/bitcoin/bitcoin/blob/11713c9fa91ca583063750cc826d0138b3a611f6/src/net.cpp#L299 as well:
```C++
            info.nScore = SaturatingAdd(nScore, is_newly_added ? 0 : 1);
```

codeabysss replied:
done thanks

l0rinc commented on src/test/net_tests.cpp:
```diff
@@ -802,6 +802,33 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+    BOOST_CHECK(SeenLocal(addr));
+    BOOST_CHECK_EQUAL(GetnScore(addr), std::numeric_limits<int>::max());
+
+    RemoveLocal(addr);
```
The existing-entry update in `AddLocal()` still computes `nScore + 1` when the supplied score is at least the stored score.

That leaves the same signed-overflow boundary reachable through the public local-address scoring helper if an existing entry is updated with `std::numeric_limits<int>::max()`.

```C++
    BOOST_CHECK(AddLocal(addr, std::numeric_limits<int>::max()));
    BOOST_CHECK_EQUAL(GetnScore(addr), std::numeric_limits<int>::max());
```

(Please make sure to update the code comments after fixing this.)

codeabysss replied:
Done added checks for AddLocal and updated the test comments. Score is read via mapLocalHost since GetnScore() is static again.

l0rinc commented on src/net.cpp:
```diff
@@ -325,7 +325,9 @@ bool SeenLocal(const CService& addr)
-    ++it->second.nScore;
+    if (it->second.nScore < std::numeric_limits<int>::max()) {
+        ++it->second.nScore;
+    }
```
We already have https://github.com/bitcoin/bitcoin/blob/8b49e2dd4eed93d08a3d9681d444aaf441ab0037/src/util/overflow.h#L45-L60 for this exact usecase, let's use it here

```suggestion
    it->second.nScore = SaturatingAdd(it->second.nScore, 1);
```

l0rinc commented on src/test/net_tests.cpp:
```diff
@@ -802,6 +802,33 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+BOOST_AUTO_TEST_CASE(LocalAddress_nScore_Overflow)
+{
+    g_reachable_nets.Add(NET_IPV4);
+    CService addr{UtilBuildAddress(0x002, 0x001, 0x001, 0x001), 1000}; // 2.1.1.1:1000
```
nit: can be `const`

l0rinc commented on src/test/net_tests.cpp:
```diff
@@ -802,6 +802,33 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+    BOOST_CHECK_EQUAL(GetnScore(addr), initial_score + 1);
+
+    // Set nScore to max.
+    RemoveLocal(addr);
```
Why is this necessarry? We could probably avoid removing and re-adding the address here.

l0rinc reviewed (changes requested):
> wrap from INT_MAX to INT_MIN triggering undefined behavior

[Signed overflow is UB](https://eel.is/c++draft/basic.fundamental#2.sentence-6), so the wrap is not guaranteed, please reword the PR description, see the  C++ draft:
> Overflow for signed arithmetic yields undefined behavior

sedited commented:
@codeabysss do you want to address the remaining comments here?

codeabysss commented:
Thanks for the review.

updates in the latest push

- Use `SaturatingAdd()` in both `AddLocal()` and `SeenLocal()`
- Reverted `GetnScore()` back to `static`
- Updated PR description to describe signed overflow as UB not a guaranteed wrap to `INT_MIN`
- Kept a focused unit test that covers saturation in both update paths, with the nits addressed `const` addr

sedited reviewed (approved):
Re-ACK 6d37cfe0eb696825760261b7f1c5ba441d7f8011

fanquake commented:
cc @maflcko

Crypt-iQ commented:
@codeabysss could you squash your 2 commits into one commit so the second commit is not changing what the first introduced?

fanquake commented:
@codeabysss are you able to squash here?

codeabysss committed:
p2p: Saturate LocalServiceInfo::nScore updates at INT_MAX

Signed overflow on nScore updates is undefined behavior. Use
SaturatingAdd in AddLocal() and SeenLocal() so increments saturate at
INT_MAX instead of overflowing.

Add unit test coverage for saturation in both code paths.

codeabysss commented:
> @codeabysss are you able to squash here?

squashed commits to one

Crypt-iQ commented on src/net.h:
```diff
@@ -102,6 +102,9 @@ static constexpr bool DEFAULT_V2_TRANSPORT{true};
 typedef int64_t NodeId;
 
+/** Get the score of a local address. */
+int GetnScore(const CService& addr);
```
Not needed anymore

codeabysss replied:
my bad forgot to clean them
its done now

Crypt-iQ commented on src/net.cpp:
```diff
@@ -222,7 +223,7 @@ CService GetLocalAddress(const CNode& peer)
 }
 
-static int GetnScore(const CService& addr)
+int GetnScore(const CService& addr)
```
Can keep `static`

codeabysss replied:
done

Crypt-iQ commented on src/test/net_tests.cpp:
```diff
@@ -802,6 +802,41 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
+    BOOST_REQUIRE(IsLocal(addr));
+    BOOST_CHECK_EQUAL(get_score(addr), initial_score);
+
+    // SeenLocal saturates nScore at INT_MAX.
```
nit: comment is wrong, this does not saturate at INT_MAX, the below just increments by 1. Can remove comment or say "SeenLocal should increment nScore by 1".

codeabysss replied:
done thanks

Crypt-iQ reviewed (commented):
Very close, left a single nit about a misleading comment

Crypt-iQ commented:
ACK 2189a6f5f226d5a2905f1939eb7eea9571502b90 pending CI

I think calling `SaturatingAdd` in `AddLocal` is good belt-and-suspenders even if it can't be hit in practice because the argument is only ever in the range `[1, 4]`.

sedited reviewed (approved):
ACK 2189a6f5f226d5a2905f1939eb7eea9571502b90

achow101 commented:
ACK 2189a6f5f226d5a2905f1939eb7eea9571502b90
